File: Code/TasmanianKoala/FRC2020Auto/src/main/java/frc/robot/utils/StopWatch.java
//----------------------------------------------------------------------------
//
//  $Workfile: StopWatch.java$
//
//  $Revision: X$
//
//  Project:    Tasmanian Koala
//
//                            Copyright (c) 2020
//                                 Jim Wright
//                            All Rights Reserved
//
//  Modification History:
//  $Log:
//  $
//
//----------------------------------------------------------------------------
//----------------------------------------------------------------------------
//  Package
//----------------------------------------------------------------------------
package frc.robot.utils;

//----------------------------------------------------------------------------
// Class Declarations
//----------------------------------------------------------------------------
//
// Class Name: StopWatch
//
// Purpose:
//  mark off an amount of time
//
//----------------------------------------------------------------------------
public class StopWatch 
{
  //----------------------------------------------------------------------------
  //  Class atributes
  //----------------------------------------------------------------------------
  long mLastTime = 0;
	int mWaitTime = 0;
	
  // --------------------------------------------------------------------
  // Purpose:
  //   Constructor
  //
  // Notes:
  //   None.
  // --------------------------------------------------------------------
  public StopWatch(int waitTime)
	{
		mWaitTime = waitTime;
	}
	
  // --------------------------------------------------------------------
  // Purpose:
  //   set the time to wait
  //
  // Notes:
  //   None.
  // --------------------------------------------------------------------
	public void setTime(int waitTime)
	{
		mWaitTime = waitTime;
	}
	
  // --------------------------------------------------------------------
  // Purpose:
  //   is the watch finished
  //
  // Notes:
  //   None.
  // --------------------------------------------------------------------
	public boolean isExpired()
	{
		if((System.currentTimeMillis()-mLastTime)>mWaitTime)
		{
			return true;
		}
		return false;
	}
	
  // --------------------------------------------------------------------
  // Purpose:
  //   Reset the watch
  //
  // Notes:
  //   None.
  // --------------------------------------------------------------------
	public void reset()
	{
		mLastTime = System.currentTimeMillis(); 
	}
	
  // --------------------------------------------------------------------
  // Purpose:
  //   return the amount of time left
  //
  // Notes:
  //   None.
  // --------------------------------------------------------------------
	public long timeLeft()
	{
		return mWaitTime - (System.currentTimeMillis()-mLastTime);
	}
}

File: Code/JetsonNano/jimscode/stopWatch.cpp
//----------------------------------------------------------------------------
//
//  $Workfile: stopWatch.cpp$
//
//  $Revision: X$
//
//  Project:    Ball Locator
//
//                            Copyright (c) 2020
//                               James A Wright
//                            All Rights Reserved
//
//  Modification History:
//  $Log:
//  $
//
//  Notes:
//     This is stopwatch counter class
//
//----------------------------------------------------------------------------

//----------------------------------------------------------------------------
//  Includes
//----------------------------------------------------------------------------
#include <stdio.h>
#include <sys/time.h>
#include "stopWatch.h"

//----------------------------------------------------------------------------
//  Purpose:
//   Constructor
//
//  Notes:
//      None
//
//----------------------------------------------------------------------------
StopWatch::StopWatch() :
  mLastTime(0),
  mWaitTime(1000)
{
}
	
//----------------------------------------------------------------------------
//  Purpose:
//   Constructor
//
//  Notes:
//      None
//
//----------------------------------------------------------------------------
StopWatch::StopWatch(int waitTime) :
  mLastTime(0),
  mWaitTime(waitTime)
{
}

//----------------------------------------------------------------------------
//  Purpose:
//   Return the current system time
//
//  Notes:
//      None
//
//----------------------------------------------------------------------------
long StopWatch::now(void)
{
  struct timeval theTime;
  gettimeofday(&theTime,NULL);
  return ((theTime.tv_sec * 1000) + (theTime.tv_usec/1000));
}

//----------------------------------------------------------------------------
//  Purpose:
//   Sets the time to wait
//
//  Notes:
//      None
//
//----------------------------------------------------------------------------
void StopWatch::setTime(int waitTime)
{
  mWaitTime = waitTime;
}

//----------------------------------------------------------------------------
//  Purpose:
//   Checks to see if the time is expired
//
//  Notes:
//      None
//
//----------------------------------------------------------------------------
bool StopWatch::isExpired(void)
{
  if((now() - mLastTime)>mWaitTime)
  {
     return true;
  }
  return false;
}

//----------------------------------------------------------------------------
//  Purpose:
//   Resets the time
//
//  Notes:
//      None
//
//----------------------------------------------------------------------------
void StopWatch::reset(void)
{
  mLastTime = now();
}

//----------------------------------------------------------------------------
//  Purpose:
//   Gets the time left to wait
//
//  Notes:
//      None
//
//----------------------------------------------------------------------------
long StopWatch::getTimeLeft()
{
  return mWaitTime - (now()-mLastTime);
}

